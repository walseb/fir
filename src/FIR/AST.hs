{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE DataKinds           #-}
{-# LANGUAGE FlexibleContexts    #-}
{-# LANGUAGE FlexibleInstances   #-}
{-# LANGUAGE GADTs               #-}
{-# LANGUAGE PolyKinds           #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications    #-}
{-# LANGUAGE TypeFamilies        #-}
{-# LANGUAGE TypeOperators       #-}

{-|
Module: FIR.AST

Representation of programs using abstract syntax trees.

The user interface to the AST is through type class overloading,
see "FIR.Instances.AST" and "FIR.Instances.Codensity".

This AST uses a higher-order abstract syntax representation (HOAS),
as seen in the paper /Combining Deep and Shallow Embeddings of Domain-Specific Languages/
by Josef Svenningsson and Emil Axelsson.
-}

module FIR.AST
  ( -- * Main AST data type
    AST(..)
    -- * Syntactic type class
  , Syntactic(Internal, toAST, fromAST)
    -- * Primitive operations
  , primOp
  , HasUndefined(undefined)
    -- * Displaying ASTs graphically
  , toTree
  )
  where

-- base
import Prelude
  hiding ( undefined )
import qualified Prelude
import Data.Kind
  ( Type )
import Data.Proxy
  ( Proxy(Proxy) )
import qualified GHC.Stack
import GHC.TypeLits
  ( Symbol, KnownSymbol, symbolVal )
import GHC.TypeNats
  ( Nat, KnownNat, natVal )

-- containers
import Data.Tree
 ( Tree(Node) )

-- mtl
import Control.Monad.State.Lazy
  ( evalState )

-- tree-view
import Data.Tree.View
  ( showTree )

-- fir
import CodeGen.Instruction
  ( ID(ID) )
import CodeGen.Monad
  ( MonadFresh(fresh), runFreshSuccT )
import Control.Type.Optic
  ( Gettable, Settable, Indices )
import Control.Monad.Indexed
  ( (:=) )
import Data.Function.Variadic
  ( NatVariadic )
import Data.Product
  ( HList, IsProduct )
import Data.Type.Known
  ( Known, knownValue )
import Data.Type.List
  ( SLength )
import FIR.ASTState
  ( TLInterface
  , ASTState(..)
  )
import FIR.Binding
  ( BindingsMap
  , FunctionType, Var, Permissions
  )
import FIR.Instances.Bindings
  ( ValidDef, ValidFunDef, FunctionTypes
  , AddBinding, AddFunBinding
  , FunctionDefinitionStartState, FunctionDefinitionEndState
  , ValidEntryPoint, SetInterface
  , GetExecutionInfo
  , EntryPointStartState, EntryPointEndState
  , Embeddable
  )
import FIR.Instances.Optics
  ( User, Assigner, Viewer, Setter, KnownOptic, SOptic, showSOptic )
import FIR.Prim.Image
  ( ImageOperands )
import FIR.Prim.Op
  ( PrimOp(PrimOpType, PrimOpConstraint, opName) )
import FIR.Prim.Singletons
  ( PrimTy, primTy, KnownVars
  , PrimFunc, primFuncName
  , KnownArity
  )
import Math.Algebra.GradedSemigroup
  ( GradedSemigroup(Grade, (:<!>:)) )
import Math.Linear
  ( V, M )
import qualified SPIRV.PrimOp as SPIRV
import qualified SPIRV.PrimTy as SPIRV
import qualified SPIRV.Stage  as SPIRV
  ( ExecutionModel, ExecutionInfo )

------------------------------------------------------------
-- main AST data type

infixl 9 :$

-- | AST representation of the EDSL.
data AST :: Type -> Type where

  -- | Lambda abstraction
  Lam :: (AST a -> AST b) -> AST (a -> b)

  -- | Function application
  (:$) :: AST (a -> b) -> AST a -> AST b

  -- | Haskell-level constants can be embedded into the AST.
  Lit :: ( PrimTy a, KnownArity a ) => a -> AST a
  -- | @SPIR-V@ primitive operations
  PrimOp :: ( PrimOp op a, PrimOpConstraint op a )
         => Proxy a -> Proxy op -> AST (PrimOpType op a)

  -- Indexed monadic operations (for the AST itself)
  -- | Indexed /return/
  Return :: AST (a -> (a := i) i)
  -- | Indexed /angelic bind/
  Bind :: AST ( (a := j) i -> (a -> q j) -> q i )

  -- | Defining a new constant/variable.
  Def :: forall k ps a i.
        ( GHC.Stack.HasCallStack
        , KnownSymbol k
        , Known Permissions ps
        , PrimTy a
        , ValidDef k i
        )
      => Proxy k  -- ^ Variable name.
      -> Proxy ps -- ^ Permissions (read,write,...).
      -> AST ( a -> (a := AddBinding k (Var ps a) i) i )

  -- | Defining a new function.
  --
  -- Meaning of type variables:
  -- * name: function name,
  -- * as: named function arguments (usually inferred),
  -- * b: function return type (usually inferred),
  -- * j_bds: bindings state at end of function definition (usually inferred),
  -- * i: monadic state at function definition site (usually inferred).
  FunDef :: forall name as b j_bds i.
            ( GHC.Stack.HasCallStack
            , KnownSymbol name
            , KnownVars as
            , PrimTy b
            , ValidFunDef name as i j_bds
            , '(as,b) ~ FunctionTypes name i
            )
         => Proxy name -- ^ Funtion name.
         -> Proxy as   -- ^ Function argument types.
         -> Proxy b    -- ^ Function return type.
         -> AST ( ( b := FunctionDefinitionEndState name as j_bds i )
                    ( FunctionDefinitionStartState name as i )
                -> ( FunctionType as b := AddFunBinding name as b i ) i
                )

  -- | Defining a new entry point.
  --
  -- An entry point is like a function definition with no arguments and Unit return type.
  -- Code within an entry point is given access to additional builtins.
  --
  -- Meaning of type variables:
  -- * name: entry point name,
  -- * stage: entry point stage,
  -- * stageInfo: entry point stage info (usually inferred),
  -- * j_bds: bindings state at end of entry point definition (usually inferred),
  -- * j_iface: entry point interface (usually inferred),
  -- * i: monadic state at entry point definition site (usually inferred),
  Entry :: forall
             ( name      :: Symbol )
             ( stage     :: SPIRV.ExecutionModel          )
             ( stageInfo :: SPIRV.ExecutionInfo Nat stage )
             ( j_bds     :: BindingsMap )
             ( j_iface   :: TLInterface )
             ( i         :: ASTState    )
           .
           ( GHC.Stack.HasCallStack
           , KnownSymbol name
           , Known SPIRV.ExecutionModel stage
           , Known (SPIRV.ExecutionInfo Nat stage) stageInfo
           , ValidEntryPoint name stageInfo i j_bds
           , stageInfo ~ GetExecutionInfo name stage i
           )
         => Proxy name      -- ^ Entry point name.
         -> Proxy stageInfo -- ^ Entry point stage info.
         -> AST (  ( () := EntryPointEndState name stageInfo j_bds j_iface i )
                   ( EntryPointStartState name stageInfo i )
                -> ( () := SetInterface name stageInfo j_iface i ) i
                )

  -- | /Use/ an optic, returning a monadic value read from the (indexed) state.
  --
  -- Like @use@ from the lens library.
  Use :: forall optic.
         ( GHC.Stack.HasCallStack, KnownOptic optic, Gettable optic )
       => SLength (Indices optic) -- ^ Singleton for the number of run-time indices.
       -> SOptic optic            -- ^ Singleton for the optic.
       -> AST ( User optic )
  -- | /Assign/ a new value with an optic.
  --
  -- Like @assign@ from the lens library.
  Assign :: forall optic.
           ( GHC.Stack.HasCallStack, KnownOptic optic, Settable optic )
         => SLength (Indices optic) -- ^ Singleton for the number of run-time indices.
         -> SOptic optic            -- ^ Singleton for the optic.
         -> AST ( Assigner optic )
  -- | /View/: access a value using an optic.
  --
  -- Like @view@ from the lens library.
  View :: forall optic.
          ( GHC.Stack.HasCallStack, KnownOptic optic, Gettable optic )
       => SLength (Indices optic) -- ^ Singleton for the number of run-time indices.
       -> SOptic optic            -- ^ Singleton for the optic.
       -> AST ( Viewer optic )
  -- | /Set/: set a value using an optic.
  --
  -- Like @set@ from the lens library.
  Set :: forall optic.
         ( GHC.Stack.HasCallStack, KnownOptic optic, Settable optic )
      => SLength (Indices optic) -- ^ Singleton for the number of run-time indices.
      -> SOptic optic            -- ^ Singleton for the optic.
      -> AST ( Setter optic )

  If    :: ( GHC.Stack.HasCallStack
           , PrimTy a
           )
        => AST ( Bool -> a -> a -> a )
  IfM   :: GHC.Stack.HasCallStack
        => AST ( Bool -> (a := j) i -> (a := k) i -> (a := i) i )
  While :: GHC.Stack.HasCallStack
        => AST ( ( Bool := i ) i -> (() := j) i -> (() := i) i )

  -- | Encapsulate local state.
  Locally :: AST ( (a := j) i -> (a := i) i )
  -- | Embed a computation into one with larger starte.
  Embed :: Embeddable i j => AST ( (a := i) i -> (a := j) j )

  -- functor/applicative operations
  Fmap :: forall f a b. ( PrimFunc f, PrimTy a, KnownArity b )
       => AST ( (a -> b) -> f a -> f b )
  Pure :: forall f a. ( PrimFunc f, KnownArity a )
       => AST ( a -> f a )
  Ap   :: forall f a b. ( PrimFunc f, PrimTy a, KnownArity b )
       => AST ( f (a -> b) -> f a -> f b )

  MkVector :: (KnownNat n, PrimTy a)
           => Proxy n
           -> Proxy a
           -> AST ( NatVariadic n a ( V n a ) )
  GradedMappend :: ( GradedSemigroup g k, a ~ Grade k g i, b ~ Grade k g j )
    => AST ( a -> b -> Grade k g (i :<!>: j) )

  -- Newtype wrapping/unwrapping.
  Mat   :: (KnownNat m, KnownNat n) => AST ( V m (V n a) -> M m n a )
  UnMat :: (KnownNat m, KnownNat n) => AST ( M m n a -> V m (V n a) )
  Ops   :: ImageOperands props ops -> AST ( ImageOperands props ops )
  Coerce :: forall a b. AST (a -> b)

  -- MkStruct
  -- MkArray

  -- | Internal HList data type.
  --
  -- Only used for providing multiple run-time indices to product optics.
  -- See [FIR issue #13](https://gitlab.com/sheaf/fir/issues/13).
  NilHList  :: AST ( HList '[] )
  ConsHList :: AST ( a -> HList as -> HList (a ': as) )
  HeadHList :: AST ( HList (a ': as) -> a )
  TailHList :: AST ( HList (a ': as) -> HList as )
  ProductToHList :: IsProduct p as => AST (p -> HList as)

  -- | Undefined.
  Undefined :: PrimTy a => AST a

  -- | As @SPIR-V@ is based around identifiers,
  -- this function can be used to create values of any type using their IDs.
  MkID :: (ID, SPIRV.PrimTy) -> AST a

------------------------------------------------

-- | Syntactic type class (Axelsson, Svenningsson).
class Syntactic a where
  type Internal a
  toAST :: a -> AST (Internal a)
  fromAST :: AST (Internal a) -> a

instance Syntactic (AST a) where
  type Internal (AST a) = a
  toAST   = id
  fromAST = id

instance (Syntactic a, Syntactic b) => Syntactic (a -> b) where
  type Internal (a -> b) = Internal a -> Internal b
  toAST        f    = Lam ( toAST . f . fromAST )
  fromAST (Lam f) a = fromAST ( f  $ toAST a )
  fromAST      f  a = fromAST ( f :$ toAST a )

------------------------------------------------

-- | Utility function for defining primops.
primOp :: forall a op r. ( PrimOp op a, PrimOpConstraint op a
                         , Syntactic r, Internal r ~ PrimOpType op a
                         )
       => r 
primOp = fromAST $ PrimOp @op @a Proxy Proxy

-- | Types at which we can define "undefined."
class HasUndefined a where
  undefined :: a

instance PrimTy a => HasUndefined (AST a) where
  undefined = Undefined

instance {-# OVERLAPPABLE #-} HasUndefined a where
  undefined = Prelude.undefined

------------------------------------------------
-- display AST for viewing

toTreeArgs :: forall m ast. MonadFresh ID m => AST ast -> [Tree String] -> m (Tree String)
toTreeArgs (f :$ a) as = do
  at <- toTreeArgs a []
  toTreeArgs f (at:as)
toTreeArgs (Lam f) as = do
  v <- fresh
  let var = MkID (v,undefined)
  body <- toTreeArgs (f var) []
  return $ case as of
    [] -> Node ("Lam " ++ show v) [body]
    _  -> Node  ":$"              (body : as)
toTreeArgs (PrimOp (_ :: Proxy a) (_ :: Proxy op) ) as 
  = return (Node ("PrimOp " ++ show ( SPIRV.op ( opName @_ @_ @op @a ) ) ) as)
toTreeArgs If        as = return (Node "If"            as)
toTreeArgs IfM       as = return (Node "IfM"           as)
toTreeArgs While     as = return (Node "While"         as)
toTreeArgs Locally   as = return (Node "Locally"       as)
toTreeArgs Embed     as = return (Node "Embed"         as)
toTreeArgs Return    as = return (Node "Return"        as)
toTreeArgs Bind      as = return (Node "Bind"          as)
toTreeArgs Mat       as = return (Node "Mat"           as)
toTreeArgs UnMat     as = return (Node "UnMat"         as)
toTreeArgs (Ops _ )  as = return (Node "ImageOperands" as)
toTreeArgs Coerce    as = return (Node "Coerce"        as)
toTreeArgs Undefined as = return (Node "Undefined"     as)
toTreeArgs NilHList  as = return (Node "NilHListAST"  as)
toTreeArgs ConsHList as = return (Node "ConsHListAST" as)
toTreeArgs HeadHList as = return (Node "HeadAST"      as)
toTreeArgs TailHList as = return (Node "TailAST"      as)
toTreeArgs ProductToHList   as = return (Node "ProductToHList"   as)
toTreeArgs (MkID     (v,_)) as = return (Node (show v) as)
toTreeArgs GradedMappend    as = return (Node "GradedMappend" as)
toTreeArgs (MkVector   n _) as = return (Node ("Vec"       ++ show (natVal n)) as)
toTreeArgs (Use    _ o    ) as = return (Node ("Use @("    ++ showSOptic o ++ ")") as)
toTreeArgs (Assign _ o    ) as = return (Node ("Assign @(" ++ showSOptic o ++ ")") as)
toTreeArgs (View   _ o    ) as = return (Node ("View @("   ++ showSOptic o ++ ")") as)
toTreeArgs (Set    _ o    ) as = return (Node ("Set @("    ++ showSOptic o ++ ")") as)
toTreeArgs (Def    k _    ) as = return (Node ("Def @"     ++ symbolVal k ) as)
toTreeArgs (FunDef k _ _  ) as = return (Node ("FunDef @"  ++ symbolVal k ) as)
toTreeArgs (Entry  _ (_ :: Proxy (stageInfo :: SPIRV.ExecutionInfo Nat stage) )) as
  = return (Node ("Entry @(" ++ show (knownValue @stage) ++ ")") as)
toTreeArgs (Lit (a :: ty)) as
  = return (Node ("Lit @(" ++ show (primTy @ty) ++ ") " ++ show a ) as)
toTreeArgs fm@Fmap as
  = case fm of
      ( _ :: AST ( (x -> y) -> f x -> f y ) )
        -> return (Node ("Fmap @(" ++ primFuncName @f ++ ") ") as)
toTreeArgs pur@Pure as
  = case pur of
      ( _ :: AST ( x -> f x ) )
        -> return (Node ("Pure @(" ++ primFuncName @f ++ ") ") as)
toTreeArgs app@Ap as
  = case app of
      ( _ :: AST ( f (x -> y) -> f x -> f y ) )
        -> return (Node ("Ap @("   ++ primFuncName @f ++ ") ") as)

toTree :: AST ast -> Tree String
toTree = (`evalState` (ID 1)) . runFreshSuccT . ( `toTreeArgs` [] )

instance Show (AST ast) where
  show = showTree . toTree
