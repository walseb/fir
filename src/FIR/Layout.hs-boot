{-# LANGUAGE AllowAmbiguousTypes   #-}
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE FlexibleContexts      #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE PolyKinds             #-}
{-# LANGUAGE ScopedTypeVariables   #-}
{-# LANGUAGE TypeApplications      #-}
{-# LANGUAGE TypeFamilies          #-}

module FIR.Layout
  ( Layout(..)
  , Poke(..)
  )
  where

-- base 
import Data.Word
  ( Word32 )
import Foreign.Ptr
  ( Ptr )
import GHC.TypeNats
  ( Nat, KnownNat )

-- fir
import Data.Type.Known
  ( knownValue )

--------------------------------------------------------------------------------------------

data Layout
  = Base
  | Extended
  | Locations

class ( KnownNat (SizeOf    lay a)
      , KnownNat (Alignment lay a)
      ) => Poke a (lay :: Layout) where

  type SizeOf lay a :: Nat
  sizeOf :: Word32
  sizeOf = knownValue @(SizeOf lay a)

  type Alignment lay a :: Nat
  alignment :: Word32
  alignment = knownValue @(Alignment lay a)

  poke :: Ptr a -> a -> IO ()
