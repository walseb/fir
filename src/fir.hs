{-# OPTIONS_HADDOCK ignore-exports #-}

{-# LANGUAGE AllowAmbiguousTypes  #-}
{-# LANGUAGE FlexibleContexts     #-}
{-# LANGUAGE FlexibleInstances    #-}
{-# LANGUAGE GADTs                #-}
{-# LANGUAGE OverloadedStrings    #-}
{-# LANGUAGE PackageImports       #-}
{-# LANGUAGE PolyKinds            #-}
{-# LANGUAGE ScopedTypeVariables  #-}
{-# LANGUAGE TemplateHaskell      #-}
{-# LANGUAGE TypeApplications     #-}
{-# LANGUAGE TypeOperators        #-}
{-# LANGUAGE UndecidableInstances #-}

{-|
Module: FIR
Description: @---@ __Main module, re-exports what's needed to write programs and compile to SPIR-V.__

This module re-exports the main functionality needed to use the library.

For an overview of the library, refer to the project __readme__.

To illustrate the syntax of this library, the example of a trivial fragment shader follows.

@
\{\-\# LANGUAGE BlockArguments   \#\-\} -- cleaner syntax for 'do' blocks (optional)
\{\-\# LANGUAGE DataKinds        \#\-\} -- for datatype promotion and type-level literals
\{\-\# LANGUAGE RebindableSyntax \#\-\} -- 'do' notation to create ASTs with indexed monads
\{\-\# LANGUAGE TypeApplications \#\-\} -- to specify type-level arguments
\{\-\# LANGUAGE TypeOperators    \#\-\} -- for type operators such as ':->, which stands for key/value assignment

import FIR
import Math.Linear -- for vectors

-- Specify the input/output of the shader, with memory locations (and other interface parameters).
-- This consists of a type-level list of top-level definitions.
type FragmentDefs
  =  '[ "in_pos"  ':-> Input      '[ Location 0      ] (V 2 Float)    -- input  (varying) of type V 2 Float and memory location 0
      , "out_col" ':-> Output     '[ Location 0      ] (V 4 Float)    -- output (varying) of type V 4 Float and memory location 0
      , "image"   ':-> Texture2D  '[ Binding  0      ] (RGBA8 UNorm)  -- input sampled image
      , "main"    ':-> EntryPoint '[ OriginLowerLeft ] Fragment       -- fragment shader stage (using standard Cartesian coordinates)
      ]

fragment :: Program FragmentDefs ()
fragment =
  Program $ entryPoint \@"main" \@Fragment do
    pos <- get \@"in_pos"
    col <- use \@(ImageTexel "image") NoOperands pos
    put \@"out_col" col
@

Note the lens-like operations:

  - @'FIR.Instances.Codensity.get' \@"in_pos"@ – equivalent to
    @'FIR.Instances.Codensity.use' \@(Name "in_pos")@ – obtains the shader input varying "in_pos".
    This operation is similar to @get@ in a state monad, except that an additional binding name
    is provided via a type application.
  - @'FIR.Instances.Codensity.use' \@(ImageTexel "image") NoOperands pos@ samples the provided image
    at coordinates @pos@. Additional operands can be provided for this sampling operation,
    such as an explicit level of detail or whether to use projective coordinates.
  - @'FIR.Instances.Codensity.put' \@"out_col" col@ – equivalent to
    @'FIR.Instances.Codensity.assign' \@(Name "out_col") col@ –
    sets the output value of the shader.


This fragment shader can then be compiled using:

> compile filePath flags fragment

where @filePath@ is the desired output filepath, and @flags@ is a list of 'CompilerFlag's.


More meaningful examples can be found in the @fir-examples@ subdirectory of the project's repository.

-}

module FIR 
  ( compile, runCompilationsTH
  , DrawableProgram(draw)
  , CompilerFlag(Debug, NoCode)
  , module Control.Monad.Indexed
  , Control.Type.Optic.Optic
  , (Control.Type.Optic.:*:)
  , (Control.Type.Optic.:.:)
  , Control.Type.Optic.Joint
  , Control.Type.Optic.AnIndex
  , Control.Type.Optic.Index
  , Control.Type.Optic.Name
  , Control.Type.Optic.ReifiedGetter(view)
  , Control.Type.Optic.ReifiedSetter(set)
  , (Data.Type.Map.:->)((:->))
  , FIR.AST.AST((:$), Lit, Pair, Ops)
  , FIR.AST.fromAST, FIR.AST.toAST -- might be a bad idea
  , FIR.Binding.BindingsMap
  , FIR.Binding.Var, FIR.Binding.Fun
  , FIR.Binding.R
  , FIR.Binding.W
  , FIR.Binding.RW
  , FIR.Definition.Definition
      ( Function
      , EntryPoint
      )
  , FIR.ASTState.ASTState(ASTState)
  , FIR.ASTState.FunctionContext(TopLevel)
  , module FIR.Instances.AST
  , module FIR.Instances.Codensity
  , module FIR.Instances.Images
  , module FIR.Instances.Optics
  , FIR.Prim.Array.Array
  , FIR.Prim.Array.mkArray
  , FIR.Prim.Array.RuntimeArray(MkRuntimeArray)
  , FIR.Prim.Image.ImageProperties(..)
  , FIR.Prim.Image.Image
  , FIR.Prim.Image.GatherInfo(..)
  , FIR.Prim.Struct.Struct(End,(:&))
  , FIR.Program.Procedure
  , FIR.Program.Program(Program)
  , module FIR.Synonyms
  , FIR.Swizzle.Swizzle
  , module Math.Algebra.Class
  , module Math.Algebra.GradedSemigroup
  , module Math.Logic.Bits
  , module Math.Logic.Class
  , SPIRV.Decoration.Decoration(..)
  , SPIRV.ExecutionMode.ExecutionMode(..)
  , SPIRV.ExecutionMode.InputVertices -- synonym of 'OutputVertices' for tessellation evaluation shaders
  , SPIRV.FunctionControl.Inlineability(..)
  , SPIRV.FunctionControl.SideEffects(..)
  , SPIRV.FunctionControl.FunctionControl
  , SPIRV.Stage.Stage(..)
  -- image properties
  , SPIRV.Image.Dimensionality(..)
  , SPIRV.Image.HasDepth(..)
  , SPIRV.Image.Arrayness(..)
  , SPIRV.Image.MultiSampling(..)
  , SPIRV.Image.ImageUsage(..)
  , SNorm, UNorm, F, I, UI
  , RGBA32, RGBA16, RGBA8
  , RG32, RG16, RG8
  , R32, R16, R8
  , R11G11B10
  , RGB10A2

  -- known
  , Data.Type.Known.Known(..)
  , Data.Type.Known.knownValue

  -- re-exporting parts of Prelude
  , Prelude.Bool(..), Prelude.otherwise
  , Prelude.Maybe(..), Prelude.maybe
  , Prelude.Either(..), Prelude.either
  , Prelude.fst, Prelude.snd
  , Prelude.curry, Prelude.uncurry
  , Prelude.Float, Prelude.Double
  , Prelude.subtract
  , Prelude.Semigroup(..), Prelude.Monoid(..)
  , Prelude.id, Prelude.const, (Prelude..)
  , Prelude.flip, (Prelude.$)
  , Prelude.until, Prelude.asTypeOf
  , Prelude.error, Prelude.errorWithoutStackTrace
  , Prelude.undefined
  , Prelude.seq, (Prelude.$!)
  , Prelude.IO
  , Prelude.FilePath

  -- numeric datatypes
  , Data.Word.Word8
  , Data.Word.Word16
  , Data.Word.Word32
  , Data.Word.Word64
  , Data.Int.Int8
  , Data.Int.Int16
  , Data.Int.Int32
  , Data.Int.Int64
  , Numeric.Half.Half

  -- overloaded strings syntax
  , Data.String.IsString(..)

  ) where

-- base
import qualified Control.Monad as Monad
  ( unless )
import Data.String
  ( IsString(fromString) )
import Prelude
import Data.Word
  ( Word8, Word16, Word32, Word64 )
import Data.Int
  ( Int8, Int16, Int32, Int64 )

-- bytestring
import qualified Data.ByteString.Lazy as ByteString

-- half
import Numeric.Half
  ( Half )

-- tree-view
import Data.Tree.View
  ( drawTree )

-- template-haskell
import qualified Language.Haskell.TH        as TH
import qualified Language.Haskell.TH.Syntax as TH

-- text-utf8
import "text-utf8" Data.Text
  ( Text )
import qualified "text-utf8" Data.Text as Text

-- fir
import CodeGen.CodeGen
import CodeGen.State
import Control.Arrow.Strength
import Control.Monad.Indexed
import Control.Type.Optic
import Data.Type.Known
import Data.Type.Map
import FIR.AST
import FIR.Binding
import FIR.Definition
import FIR.Instances.AST
import FIR.Instances.Codensity
import FIR.Instances.Images
import FIR.Instances.Optics
import FIR.ASTState
import FIR.Prim.Array
import FIR.Prim.Image
import FIR.Prim.Struct
import FIR.Program
import FIR.Synonyms
  hiding
    ( Row_, Col__, Ix_, Ix__
    , RowRes, ColRes, DiagRes, EntryRes
    ) -- internal helpers
import FIR.Swizzle
import Math.Algebra.Class
import Math.Algebra.GradedSemigroup
import Math.Logic.Bits
import Math.Logic.Class
import SPIRV.Decoration
import SPIRV.ExecutionMode
import SPIRV.FunctionControl
import SPIRV.Image
import SPIRV.Stage

------------------------------------------------

-- | Functionality for drawing a program AST as a tree.
--
-- Can be used on whole programs, or on snippets of code in the AST indexed monad.
class DrawableProgram prog where
  draw :: prog -> IO ()

instance DrawableProgram (AST a) where
  draw = drawTree . toTree
instance DrawableProgram (Codensity AST (AST a := j) i) where
  draw = drawTree . toTree . toAST
instance ( DrawableProgram
             ( CodensityProgram (StartState defs) (EndState defs) a )
         )
      => DrawableProgram (Program defs a)
      where
  draw (Program prog) = draw prog

-- | Compiler flags.
data CompilerFlag
  = NoCode -- ^ Don't emit any SPIR-V code.
  | Debug  -- ^ Include additional debug instructions, such as source-code line-number annotations.
  deriving ( Prelude.Eq, Show )

-- | Compiles a program, saving the SPIR-V assembly at the given filepath.
compile :: forall defs a. KnownDefinitions defs
        => FilePath
        -> [CompilerFlag]
        -> Program defs a 
        -> IO ( Either Text Text )
compile filePath flags (Program program) = case runCodeGen cgContext (toAST program) of
    Left  err -> Prelude.pure ( Left err )
    Right bin
      ->  do  Monad.unless ( NoCode `elem` flags )
                ( ByteString.writeFile filePath bin )
              Prelude.pure ( Right "OK" )
  where cgContext :: CGContext
        cgContext = (initialCGContext @defs) { debugMode = Debug `elem` flags }

instance TH.Lift Text where
  lift t = [| Text.pack $(TH.lift $ Text.unpack t) |]

-- | Utility function to run IO actions at compile time using Template Haskell.
-- Useful for compiling shaders at compile-time, before launching a graphics application.
--
-- __Usage__: in a module which imports this module and shaders @vertexShader@, @fragmentShader@
-- (bearing in mind the TH staging restriction), write:
--
-- > shaderCompilationResult =
-- >   $( runCompilationsTH
-- >        [ ("Vertex shader"  , compile vertPath [] vertexShader  )
-- >        , ("Fragment shader", compile fragPath [] fragmentShader)
-- >        ]
-- >   )
--
-- This will compile the vertexShader @vertexShader@ at filepath @vertPath@,
-- and similarly for the fragment shader.
-- These can then be loaded into Vulkan shader modules for use in rendering.
runCompilationsTH :: [ ( Text, IO (Either Text Text) ) ] -> TH.Q TH.Exp
runCompilationsTH namedCompilations
  = TH.lift Prelude.=<< TH.runIO (combineCompilations namedCompilations)
    where
      combineCompilations :: [ ( Text, IO (Either Text Text) ) ] -> IO (Either Text Text)
      combineCompilations
        = fmap ( foldl ( \ b (n,a) -> combineResult n b a ) (Right "OK") )
        . traverse ( uncurry rightStrength )

      combineResult :: Text -> Either Text Text -> Either Text Text -> Either Text Text
      combineResult _    a           (Right _ ) = a
      combineResult name (Right _)   (Left err) = Left (name <> ": " <> err)
      combineResult name (Left errs) (Left err)
        = Left (errs <> "\n" <> name <> ": " <> err)
