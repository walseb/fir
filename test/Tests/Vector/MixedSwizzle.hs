{-# LANGUAGE AllowAmbiguousTypes   #-}
{-# LANGUAGE BlockArguments        #-}
{-# LANGUAGE DataKinds             #-}
{-# LANGUAGE FlexibleInstances     #-}
{-# LANGUAGE GADTs                 #-}
{-# LANGUAGE NamedWildCards        #-}
{-# LANGUAGE OverloadedLabels      #-}
{-# LANGUAGE PatternSynonyms       #-}
{-# LANGUAGE PartialTypeSignatures #-}
{-# LANGUAGE PolyKinds             #-}
{-# LANGUAGE RebindableSyntax      #-}
{-# LANGUAGE ScopedTypeVariables   #-}
{-# LANGUAGE TypeApplications      #-}
{-# LANGUAGE TypeFamilies          #-}
{-# LANGUAGE TypeOperators         #-}
{-# LANGUAGE UndecidableInstances  #-}

module Tests.Vector.MixedSwizzle where

-- fir
import FIR
import Math.Linear

---------------------------------------------------------------------------------

type Defs = '[ "in_col"  ':-> Input      '[Location 0] (V 4 Float)
             , "out_col" ':-> Output     '[Location 0] (V 4 Float)
             , "main"    ':-> EntryPoint '[OriginUpperLeft] Fragment
             ]


program :: Module Defs ()
program = Module $ entryPoint @"main" @Fragment do

    col <- use @(Name "in_col" :.: Swizzle "rgxb") -- cannot mix "rgba" and "xyzw"

    put @"out_col" col
