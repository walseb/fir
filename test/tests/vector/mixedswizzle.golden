
test\tests\vector\mixedswizzle.hs:35:12: error:
    • Cannot mix and match between different swizzle sets.
      Swizzle set of "g" is "rgba".
      Swizzle set of "x" is "xyzw".
    • In a stmt of a 'do' block:
        col <- use @(Name "in_col" :.: Swizzle "rgx")
      In the third argument of ‘entryPoint’, namely
        ‘do col <- use @(Name "in_col" :.: Swizzle "rgx")
            put @"out_col" col’
      In the second argument of ‘($)’, namely
        ‘entryPoint
           @"main"
           @Fragment
           do col <- use @(Name "in_col" :.: Swizzle "rgx")
              put @"out_col" col’
   |
35 |     col <- use @(Name "in_col" :.: Swizzle "rgx") -- cannot mix "rgba" and "xyzw"
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

test\tests\vector\mixedswizzle.hs:35:12: error:
    • Cannot mix and match between different swizzle sets.
      Swizzle set of "g" is "rgba".
      Swizzle set of "x" is "xyzw".
    • In a stmt of a 'do' block:
        col <- use @(Name "in_col" :.: Swizzle "rgx")
      In the third argument of ‘entryPoint’, namely
        ‘do col <- use @(Name "in_col" :.: Swizzle "rgx")
            put @"out_col" col’
      In the second argument of ‘($)’, namely
        ‘entryPoint
           @"main"
           @Fragment
           do col <- use @(Name "in_col" :.: Swizzle "rgx")
              put @"out_col" col’
   |
35 |     col <- use @(Name "in_col" :.: Swizzle "rgx") -- cannot mix "rgba" and "xyzw"
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

test\tests\vector\mixedswizzle.hs:35:12: error:
    • Cannot mix and match between different swizzle sets.
      Swizzle set of "g" is "rgba".
      Swizzle set of "x" is "xyzw".
    • In a stmt of a 'do' block:
        col <- use @(Name "in_col" :.: Swizzle "rgx")
      In the third argument of ‘entryPoint’, namely
        ‘do col <- use @(Name "in_col" :.: Swizzle "rgx")
            put @"out_col" col’
      In the second argument of ‘($)’, namely
        ‘entryPoint
           @"main"
           @Fragment
           do col <- use @(Name "in_col" :.: Swizzle "rgx")
              put @"out_col" col’
   |
35 |     col <- use @(Name "in_col" :.: Swizzle "rgx") -- cannot mix "rgba" and "xyzw"
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
